livre/lib/pipeline/HistogramFilter.cpp

```diff
@@ -0,0 +1,178 @@
+        const bool isNegYBorder = border.y() == 0.0f;
+        const bool isPosYBorder = border.y() + border.w() == 1.0f;
+
+        if( isNegXBorder )
+            minBox[ 0 ] = -infinite;
+        if( isPosXBorder )
+            maxBox[ 0 ] = infinite;
+        if( isNegYBorder )
+            minBox[ 1 ] = -infinite;
+        if( isPosYBorder )
+            maxBox[ 1 ] = infinite;
+    }
+
+    void execute( const FutureMap& input, PromiseMap& output ) const
+    {
+        const auto frustum = input.get< Frustum >( "Frustum" ).front();
+        const auto viewport = input.get< Viewport >( "RelativeViewport" ).front();
+
+        Histogram histogramAccumulated;
+        for( const auto& cacheObjects: input.getFutures( "CacheObjects" ))
+            for( const auto& cacheObject: cacheObjects.get< ConstCacheObjects >( ))
+            {
+                const CacheId& cacheId = cacheObject->getId();
+                ConstCacheObjectPtr histCacheObject = _histogramCache.load( cacheId );
+                if( !histCacheObject->isLoaded( ))
+                    continue;
+
+                const ConstHistogramObjectPtr& histogramObj =
+                        std::static_pointer_cast< const HistogramObject >(
+                            histCacheObject );
+
+                // When a frame is rendered in multi-channel, multi-node, etc
+                // config, some nodes are rendered twice in sort-first renderings
+                // To avoid that, we check whether the center of the node is in
+                // frustum or, if node is in the viewport border. To do this we
+                // project the center of the node to the normalized device coordinates
+                // (NDC), in case of partial intersection we extend the borders of
+                // NDC in border directions
+                const LODNode& lodNode = _dataSource.getNode( NodeId( cacheId ));
+                if( isCenterInViewport( frustum, lodNode.getWorldBox( )))
+                {
+                    histogramAccumulated += histogramObj->getHistogram();
+                    continue;
+                }
+
+                const vmml::Visibility result = frustum.getBoxVisibility( lodNode.getWorldBox( ));
+                switch( result )
+                {
+                    case vmml::VISIBILITY_PARTIAL:
+                    {
+                        Vector3f minBox( -1.0f, -1.0f, -infinite );
+                        Vector3f maxBox( 1.0f, 1.0f, infinite );
+                        scaleNdcIfBorder( viewport, minBox, maxBox );
+                        if( isCenterInViewport( frustum, lodNode.getWorldBox( ), minBox, maxBox ))
+                            histogramAccumulated += histogramObj->getHistogram();
+                    }
+                    break;
+                    case vmml::VISIBILITY_FULL:
+                    case vmml::VISIBILITY_NONE:
+                    default:
+                        break;
+                }
+            }
```

Comment:
I think you can remove the check for frustum.getBoxVisibility and simply replace lines 103-124 by this:

    Vector3f minBox( -1.0f, -1.0f, -infinite );
    Vector3f maxBox( 1.0f, 1.0f, infinite );
    scaleNdcIfBorder( viewport, minBox, maxBox );
    if( isCenterInViewport( frustum, lodNode.getWorldBox( ), minBox, maxBox ))
        histogramAccumulated += histogramObj->getHistogram();